R/ard_stack_hierarchical.R

```diff
@@ -0,0 +1,403 @@
+  check_not_missing(variables)
+  # denominator must be empty, a data frame, or integer
+  if (!is_empty(denominator) && !is.data.frame(denominator) && !is_integerish(denominator)) {
+    cli::cli_abort(
+      "The {.arg denominator} argument must be empty, a {.cls data.frame}, or an {.cls integer}, not {.obj_type_friendly {denominator}}.",
+      call = get_cli_abort_call()
+    )
+  }
+
+  # create ARD -----------------------------------------------------------------
+  internal_stack_hierarchical(
+    data = data,
+    variables = {{ variables }},
+    by = {{ by }},
+    id = NULL,
+    denominator = denominator,
+    include = {{ include }},
+    overall = overall,
+    overall_row = overall_row,
+    attributes = attributes,
+    total_n = total_n,
+    shuffle = shuffle
+  )
+}
+
+
+
+internal_stack_hierarchical <- function(data,
+                                        variables,
+                                        by = dplyr::group_vars(data),
+                                        id = NULL,
+                                        denominator = NULL,
+                                        include = everything(),
+                                        overall = FALSE,
+                                        overall_row = FALSE,
+                                        attributes = FALSE,
+                                        total_n = FALSE,
+                                        shuffle = FALSE) {
+  # process inputs -------------------------------------------------------------
+  check_not_missing(data)
+  check_not_missing(variables)
+  cards::process_selectors(data, variables = {{ variables }}, id = {{ id }}, by = {{ by }})
+  cards::process_selectors(data[variables], include = {{ include }})
+  check_scalar_logical(overall)
+  check_scalar_logical(overall_row)
+  check_scalar_logical(attributes)
+  check_scalar_logical(total_n)
+  check_scalar_logical(shuffle)
+
+  # check inputs ---------------------------------------------------------------
+  # both variables and include must be specified
+  if (is_empty(variables) || is_empty(include)) {
+    cli::cli_abort(
+      "Arguments {.arg variables} and {.arg include} cannot be empty.",
+      call = get_cli_abort_call()
+    )
+  }
+
+  # the last `variables` variable should be included
+  if (!utils::tail(variables, 1L) %in% include) {
+    cli::cli_abort(
+      "The last column specified in the {.arg variables} (i.e. {.val {utils::tail(variables, 1L)}}) must be in the {.arg include} argument.",
+      call = get_cli_abort_call()
+    )
+  }
+
+  if (is_empty(by) && isTRUE(overall)) {
+    cli::cli_inform(
+      c("The {.arg by} argument must be specified when using {.code overall=TRUE}.",
+        i = "Setting {.code overall=FALSE}.")
+    )
+    overall <- FALSE
+  }
+
+  if (!is.data.frame(denominator) && isTRUE(overall)) {
+    cli::cli_inform(
+      c("The {.arg denominator} argument must be specified with a data frame when using {.code overall=TRUE}.",
```

Comment:
OH boy, I have to think about this. hahaa

You can specify denom for the counts function, but it's hard to imagine a _reasonable way_ to do it with a `by` variable. I think this condition should start with `!is_empty(denominator)` and that should solve the issue.

Reply:
OK OK, I think this is actually correct, because when we run the overall results, we define the `by` variable with `by = setdiff(by, names(denominator))`. In _most_ cases, `by=NULL` for overall. BUT when `by = c(TRTA, AESEV)`, our overall results should be calculated with `by = AESEV`, and to know exactly which columns to remove, we need to know which also appear in `denominator`.

Anyway, I think this is the preferred behavior. But given that it was hard to remember only 24 hours after writing it....I am concerned this is a bad way to go about it.

Reply:
I've updated the description of this argument to be accurate and also added a detailed example in a new section called `Overall Argument`